source combinators: extend, trace, cutAt, focusAt and more

roberth opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

_Improve usability of `lib.sources`, specifically_

 - Extend the capabilities of `lib.sources`, primarily:
   - `extend` to construct sources from pre-existing sources in related directories
   - allow a source to point to a subpath inside the generated store path
 - Add documentation.
 - Add tests.

Refactor `cleanSourceWith` to operate on a simple internal representation, in preparation for new combinators to be added later.

I've discussed a union-like combinator [here before.](https://github.com/NixOS/nixpkgs/pull/56985#pullrequestreview-551210415)

`extend` combines naturally with the [`subDir`](https://github.com/input-output-hk/haskell.nix/blob/master/lib/clean-source-with.nix#L34-L36) feature in the `haskell.nix` library. This became `subpath`. `includeSiblings` was replaced by `cutAt` and/or `extend`.

###### Things done


<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] nix-build lib/tests/release.nix, also run by ofborg ~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Infinisil commented:
Sounds exciting! I guess you already have an idea for how to design combinators? I'm sure @Profpatsch is going to be interested too. Also pinging @nh2 as the author of #56985

roberth commented:
@Infinisil thanks. I've added my ideas to the description for context, but this PR can be reviewed independently from those.

roberth committed:
lib/tests/property-test.nix: init

roberth committed:
lib.filesystem.pathHasPrefix: init

roberth committed:
lib.filesystem.{commonPath,absolutePathComponentsBetween}: init

roberth committed:
lib.sources.setName: init

This isn't very new, but makes the functionality discoverable and
fits the new naming scheme.

roberth committed:
lib.sources.union: init

roberth commented:
`unionSource` got a bit more complicated than I expected because of the "subpath invariant" as described in the inline comments.

roberth commented:
I have an idea that may improve the feel of this interface. The new functions have a new style of interface that is a bit distinct from `cleanSourceWith`.

`cleanSourceWith`:
  - has `Source` in the name
  - params are an attribute set, doing many things at once. Too many?
  - users can't find the `rename` function because it's a `name` parameter in `cleanSourceWith`
  - looks unappealing when mixed with the other combinators like `unionSource` or `reparentSource`

```nix
cleanSourceWith { name = "myproject-src"; src = union ./pom.xml (cleanSource ./src); }
```

`source.filter` (idea):
 - uses `source` attrset for namespacing
 - params are minimalistic: just the filter function and original source. If you need something else, use a different function
 - matches the "style" of `source.union`
 - can be used with `inherit` or `with`

```nix
with sources;
rename "myproject-src" (union ./pom.xml (clean ./src))
```

I think I've just convinced myself.

roberth committed:
lib.sources.filter: init

roberth committed:
lib.sources.extend: init

roberth committed:
lib.sources: Subpaths and extend with parent files

roberth committed:
WIP stdenv.mkDerivation: Support sources with subpath

roberth committed:
lib.sources.union: Make private

It does not match the expectations that one may have from a
function named union; specifically, it is not associative due
to handling of the source root.

sources.extend is a better alternative, which does not seem
to imply such a property.

roberth committed:
lib.sources.cutAt: init

roberth commented:
I went ahead and implemented the subpath feature. You can now write a derivation that uses files from parent directories. For example, you may want to write a Nix expression in a project's `doc` directory to generate html.

 - `makelib/lib.make` with makefile rules
 - `README.md`; another file we want to use
 - `doc/Makefile` that has `include ../makelib/lib.make`

It's a bit of a contrived example, but situations like these happen all the time with various auxiliary files aren't directly in the subpath where you need them. With this PR, we don't have to touch the existing makefiles and we don't have to base our source on the parent directory and filter away almost everything. `doc/default.nix` can look like:

```nix
{ stdenv, pandoc, lib }:
stdenv.mkDerivation {
  name = "README.html";
  nativeBuildInputs = [ pandoc ];
  src = with lib.sources; extend (cleanSource ./.) [(cleanSource ../makelib) ../README.md];
}
```

roberth committed:
lib.sources: Generate docs and sanitize source name

roberth committed:
lib.sources: Rewrite extend to be more efficient

Avoids calling filter functions that don't apply to a subdir
because the subdir itself got false. Reduces the memoization to
one tree per extend call instead of one tree for each extend
source.

This commit gets rid of the internal `union` definition as well.
I've thought about reintroducing a `union` function on lists,
but it would require a non-empty list, because its path needs
to be defined.
Supporting special "empty" source objects without a path
would complicate the implementation unnecessarily.

davidscherer commented:
I stumbled across this PR because I was looking for exactly this.  I am prototyping using Nix as a build system for C/C++ code (doing incremental builds by generating one derivation per object file using IFD, without make) and these combinators are an excellent fit for specifying source and include paths and then combining them.  Thanks for this work!

A few comments:

1. Missing paths don't seem to always generate errors when they should.  `extend (empty ./.) [ ./does-not-exist ]` gives an empty directory rather than an error, and I think it's important that it does.
2. `extend` claims to accept SourceLike base but actually only accepts a Source.  For example `extend ./foo [ ./bar ]` gives "value is a path while a set was expected", but `extend (cleanSourceWith { src=./foo; }) [ ./bar ]` works as expected.  I think this is just because of a couple of stray references to `base` that should be `baseAttrs`.  I didn't audit other combinators for the same issue.
3. `exclude` with the same type signature as extend but subtracting the sets of files would be very useful (for excluding individual directories and files; for excluding files by suffix; for inverting regular expressions).  At the moment I think the only recourse for these is `filter`
4. Maybe `extend` should keep the name of the base?
5. I am not sure it is worth changing, but the lack of clean "top" and "bottom" members of Source and the consequent inability to union or intersect empty lists does hurt a little.  For example, below you will see an example where I map sourceFilesBySuffices over a (possibly empty) array of SourceLikes; the code isn't bad but it would be more elegant to apply sourceFilesBySuffices to the union/extension of all the sourcelikes.  I suppose you could make it so that union [] is a special "bottom" lacking a root or subpath, and make it an error to try to get an outPath for such a Source without first unioning it with something else.
6. I now want similar combinators for other things like link farms!
7. In order to make my use case work, I had to look at .origSrc and .subpath (easy, but I get the impression you prefer it to be encapsulated), and I had to write `getRelativePathFrom` (basically absolutePathComponentsBetween but willing to generate `..` components if its parameters already share a common root).  What I'm doing is probably a little different from the use cases you imagined for sources, but I will go into a little detail in case it gives you any ideas.  (Alternatively, you might just want to embrace Hyrum's Law and live with the idea that all the members of the Source type are interface.)

As I said, I'm attempting to build C++ source directly using Nix, replacing a Makefile, with the ability to rebuild incrementally (but with a fully reproducible result) when only some of the source has changed.  So you can do something like this:

```nix
c2nix.buildCBinary {
    src = ./.;
    include_src = with lib.sources; [
        (empty ../.)  # The parent directory is an include directory, but no sources
        ./.
        ../common/include
        (extend (empty ../other_subsystem) [
            ../other_subsystem/shared_header.h
            ../other_subsystem/other_header.h
        ])
    ];
}
```

The entries in the include_src array are both grabbing source files and specifying the include path that will be handed to the compiler.  This is easier, less error prone, and more flexible than without sources.

Internally buildCBinary then further uses your combinators to filter and combine these into a single source directory:

```nix
extend
    (sourceFilesBySuffices src [ ".c" ".cpp" ".h" ".hpp" ".cc" ])
    (map (s: sourceFilesBySuffices s [".h" ".hpp"]) include_src);
```

So far I think this is all in line with your intended vision.  But in order to extract an include path from include_src, I need `getRelativePathFrom`:

```nix
all_include_dirs = lib.lists.unique
                ( map ( getRelativePathFrom all_src ) include_src );

# Returns a relative path such that path == src + (getRelativePathFrom src path)
# throws an error (via absolutePathComponentsBetween) if the *root* of path is not a prefix of src,
#   but getRelativePathFrom will generate ../ components if they are already siblings in the same root
#   (e.g. because src has already been extended with path)
getRelativePathFrom = src: path: 
    let
        clean_src = lib.sources.cleanSourceWith { src=src; };
        clean_path = lib.sources.cleanSourceWith { src=path; };
        relative_to_root = (lib.filesystem.absolutePathComponentsBetween clean_src.origSrc clean_path.origSrc) ++ clean_path.subpath;
        get_relative = a: b:
            if a == [] then b
            else if b != [] && builtins.head a == builtins.head b then get_relative (builtins.tail a) (builtins.tail b)
            else (map (_:"..") a) ++ b;
        relative_to_src = get_relative clean_src.subpath relative_to_root;
    in
        if relative_to_src!=[] then (lib.concatStringsSep "/" relative_to_src) else ".";
```

and maybe something like that should be in lib since it looks at .origSrc and .subpath.

Moreover, as I said I'm providing incremental builds.  I run the C preprocessor over each module in the source to produce dependency information files (at evaluation time!), then read those (using Import From Derivation) to generate one derivation per object file having just the dependencies it needs.  I can't give those derivations dependencies on the source path, because then everything will rebuild when any source file changes (instead of just the preprocessor step rebuilding, which fortunately is orders of magnitude faster).  Therefore I need to extract `all_src.origSrc` and `all_src.subpath` in order to convert relative paths in the dependency information (referring to files in the source outPath) to direct references to the original source files.

I'm not sure if there is something that could be added to the library to enable this that wouldn't boil down to exposing these.  (Of course, mere getters would still preserve some ability to change representations.)  A `getOriginalSourcePath = srclike: let src=toSourceAttributes srclike; in src.origSrc + (lib.concatStringsSep "/" ([""] ++ src.subpath))` together with getRelativePathFrom would serve my needs for .origSrc, but I would still need to look at subpath for similar reasons to why stdenv does.

roberth committed:
lib.sources: Always validate paths

roberth committed:
lib.sources: Use naming convention to avoid bug

The raw sourcelike must not be used.

roberth committed:
lib.sources.{cutAt,pointAt}: assert type of path argument

roberth committed:
lib.sources: Support paths with metadata

roberth committed:
lib.sources.extend: Use name from first argument

roberth committed:
lib.sources.{pointAt -> focusAt}

A less ambiguous name that can also be used as a noun.

Also who doesn't love focus these days?

roberth committed:
lib.sources.getOriginalFocusPath: init

roberth committed:
lib.sources.getSubpath: init

roberth commented:
@davidscherer That's an interesting project! It seems to be similar to [snack](https://github.com/nmattia/snack), although I'm not too familiar with that. I believe these solutions can become mainstream (in the Nix ecosystem) when they run inside [ret-cont recursive nix](https://github.com/NixOS/rfcs/pull/40), which solves the evaluation performance problem. Flakes may alleviate this, but not completely because its caching is too coarse. (getting a bit off topic; can switch to discourse to keep this thread focused)

Anyway, I'm glad you've found this PR!

> 1. Missing paths don't seem to always generate errors when they should.  `extend (empty ./.) [ ./does-not-exist ]` gives an empty directory rather than an error, and I think it's important that it fails.

Fixed generically in `toSourceAttributes`.

> 2. `extend` claims to accept SourceLike base but actually only accepts a Source.  For example `extend ./foo [ ./bar ]` gives "value is a path while a set was expected", but `extend (cleanSourceWith { src=./foo; }) [ ./bar ]` works as expected.  I think this is just because of a couple of stray references to `base` that should be `baseAttrs`.  I didn't audit other combinators for the same issue.

Fixed by using a `*Raw` suffix in a couple of places, making this implementation error harder to make.

> 3. `exclude` with the same type signature as extend but subtracting the sets of files would be very useful (for excluding individual directories and files; for excluding files by suffix; for inverting regular expressions).  At the moment I think the only recourse for these is `filter`

I fully agree that this would be useful, but I'm hesitant to implement it as part of this PR. `extend` was already a surprisingly deep rabbit hole.

> 4. Maybe `extend` should keep the name of the base?

Done. It wasn't appropriate for a `union`-like operation, but for `extend` it is.

> 5. I am not sure it is worth changing, but the lack of clean "top" and "bottom" members of Source and the consequent inability to union or intersect empty lists does hurt a little.  For example, below you will see an example where I map sourceFilesBySuffices over a (possibly empty) array of SourceLikes; the code isn't bad but it would be more elegant to apply sourceFilesBySuffices to the union/extension of all the sourcelikes.  I suppose you could make it so that union [] is a special "bottom" lacking a root or subpath, and make it an error to try to get an outPath for such a Source without first unioning it with something else.

Yeah, I've struggled with this. Ideally we'd only have algebraic combinators: a family of functions that all accept the same types they return. `union` can not behave like that for `[]`, while it has the name of a widely known algebraic operator. Indeed we could extend the type to allow this kind of failure, but that doesn't seem beneficial because of the apparent dilemma of how `outPath` should behave when it doesn't really exist. If we make it `null`, that value will propagate to functions that do not expect it and can not provide a helpful error message. If we make it throw an exception, it will be hard to consume it in cases where it can be handled. Exceptions can be caught, but it's better to use an extra attribute so we know in advance what is up. At this point we're adding a lot of complexity though, which I'm not sure whether it's worth it. Perhaps it's best to explore this in a follow-up PR.

For now, do you think it's feasible for you to use lists of sources in more places? Perhaps more places than you'd feel inclined to use them.

> 6. I now want similar combinators for other things like link farms!

Yes, there's probably something to be gained there. I don't know what a design for those would look like. Link farms seem to be more complicated than sources. I hope I'm wrong though :).

> I get the impression you prefer it to be encapsulated

Kind of. We usually can't hide things in Nix; only overcomplicate them. It's not high on my list of priorities.

> 7. In order to make my use case work, I had to look at .origSrc and .subpath (), and I had to write `getRelativePathFrom` [...].  What I'm doing is probably a little different from the use cases you imagined for sources, but I will go into a little detail in case it gives you any ideas.  (Alternatively, you might just want to embrace Hyrum's Law and live with the idea that all the members of the Source type are interface.)

Hyrum's law is spot on here, but I'd like to improve the experience by providing useful functions where it makes sense.

>     include_src = with lib.sources; [

Not sure if I should read much into this, but is this a list because it's a "nullable" source (as I also suggested in this comment) or for some other reason. Also, the naming convention is camel case, with exceptions for package names and other things that map to externally defined identifiers, such as RFC42 settings attributes.

>     (sourceFilesBySuffices src [ ".c" ".cpp" ".h" ".hpp" ".cc" ])

Do you accept `cxx`, `hxx`?

> `getRelativePathFrom`

This function kind of exists at both the `filesystem.nix` and `sources.nix` layers. The implementation seems a bit complicated. Did you see `commonPath`?

> and maybe something like that should be in lib since it looks at .origSrc and .subpath.

And it's generally useful.

> I'm not sure if there is something that could be added to the library to enable this that wouldn't boil down to exposing these. (Of course, mere getters would still preserve some ability to change representations.) A `getOriginalSourcePath = srclike: let src=toSourceAttributes srclike; in src.origSrc + (lib.concatStringsSep "/" ([""] ++ src.subpath))` together with getRelativePathFrom would serve my needs for .origSrc, but I would still need to look at subpath for similar reasons to why stdenv does.

I've added `getOriginalFocusPath` and `getSubpath`.
A function for resolving relative paths would be a good addition but I don't know if it's going to be part of this PR. The line count is already massive.

I've renamed `pointAt` to `focusAt`. It seems less ambiguous and it can be used as a noun or adjective which was helpful for `getOriginalFocusPath`.

roberth commented:
@hamishmack The subpath/focus feature introduced here should be equivalent to [`subDir`](https://github.com/input-output-hk/haskell.nix/blob/master/lib/clean-source-with.nix#L34-L36). Would you be able to review this PR from the perspective of `haskell.nix`?

davidscherer commented:
I've tried the latest commit, and didn't notice any regressions in my stuff.

> ret-cont recursive nix

Interesting.  Yes, I can get significant parallelism *within* the preprocessing build step by preprocessing each module in parallel and so the evaluation performance doesn't *seem* like a fatal problem for building my team's own software (I'm not far enough along to even be sure of that), but I suspect you would need more parallelism if you suddenly had the upstream of everything in nixpkgs wanting to build this way.  I think it makes sense for Nix to support dynamic build graphs more efficiently in one way or another.

(I've thought about using some kind of metadata to identify derivations that will be used for IFD, so an explicit pre-build step can get them built eagerly to avoid blocking evaluation... but that is obviously a limited solution)

> Indeed we could extend the type to allow this kind of failure, but that doesn't seem beneficial because of the apparent dilemma of how outPath should behave when it doesn't really exist

I would vote for outPath throwing - the idea is that such Sources are only permitted as parameters to certain other combinators, and should eagerly die if they are seen anywhere else.

Maybe you could go even farther than top and bottom, and be able to implement `not` (of course the complement of any normal source can't be put in the store, but it can presumably be intersected).  I'm speculating here, though: what you have now is very useful and I personally want it merged so I can use it.

> For now, do you think it's feasible for you to use lists of sources in more places? Perhaps more places than you'd feel inclined to use them.

Yes, I've been able to work around this fine so far.

> is this a list because it's a "nullable" source (as I also suggested in this comment) or for some other reason.

The main reason it's a list is not for optionality but because the focus (+1 for a term that can be a noun, my own documentation just got less awkward) of each source in the list becomes an include directory in that all_include_dirs calculation.  From the user perspective, if you pass `includeSrc = [ ../foo ../bar ]` then you can `#include "foo.h"` instead of `#include "../foo/foo.h"`.  The focus mechanism allows you to manipulate this if you need to (usually with `reparent`), without requiring the list of include directories to (frequently) be duplicated in a difficult to verify way, but there needs to be more than one focus, so it makes sense to pass a list rather than a union.

> Also, the naming convention is camel case

oops, my brain no longer even sees the different ways of escaping word boundaries in identifiers

> cxx, hxx

Sure, why not

> getOriginalFocusPath and getSubpath

With these my use of the attributes of Source seems to be reduced to getRelativePathFrom.  

> [getRelativePathFrom] kind of exists at both the filesystem.nix and sources.nix layers

The reason I would prefer to have something like it in sources (as opposed to filesystem or elsewhere) is because its ability to throw an error if the path isn't inside the source's root seems like an important safeguard.  A general purpose relative path generator for paths would happily generate relative paths navigating between nix store paths, out of the nix store and into your home directory and vice versa... it seems like a footgun even if it generates the exact same result in every non-error case.

> The implementation seems a bit complicated. Did you see commonPath?

I certainly can't claim my implementation is optimal or even that I am aware of all the library functions I could be using.  If I had seen a removeCommonPath that returned the *tail* of each path, I might have tried to use it, but even then converting to strings early makes things more awkward for generating `..` components.  Feel free to make me look dumb by supplying a simple implementation that is equivalent :-)

wwilson commented:
What are the steps to getting this reviewed and merged? I would love to be able to use this functionality.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/source-combinators/13331/1

roberth commented:
You could use it in a flake as

```nix
  inputs.source-combinators.url = "github:hercules-ci/nixpkgs/source-combinators";

  outputs = { self, nixpkgs, source-combinators }:
  let
    inherit (source-combinators.lib) sources;
  in
    { foo = f sources.extend; # etc, etc
```

in doing so, you can provide feedback and validate the new interface.

Or quick and dirty, non-reproducible, without flakes:

```nix
let
  inherit (import (builtins.fetchTarball "https://github.com/hercules-ci/nixpkgs/archive/refs/heads/source-combinators.tar.gz" + "/lib")) sources
in
  sources.extend # etc, etc
```

ldesgoui commented on lib/tests/sources.sh:
```diff
@@ -0,0 +1,322 @@
+    ) == "src"
+  )}")'
+)"
+[[ true == "$bool" ]] || die "getSubpath example 1"
```
```suggestion
[[ true == "$string" ]] || die "getSubpath example 1"
```
Seems like a typo here?

roberth replied:
Fixed. Thanks!

ldesgoui commented on lib/tests/sources.sh:
```diff
@@ -0,0 +1,322 @@
+    ) == ""
+  )}")'
+)"
+[[ true == "$bool" ]] || die "getSubpath example 2"
```
```suggestion
[[ true == "$string" ]] || die "getSubpath example 2"
```

roberth commented:
I've opened https://github.com/NixOS/nixpkgs/pull/124875 which has a bite-sized portion of this PR.

@wwilson You could use this in a flake with

```
inputs.source-combinators.url = "github:hercules-ci/nixpkgs/source-combinators";
outputs = { self, nixpkgs, source-combinators }:
  let
    inherit (source-combinators.lib) sources;
  in
```

which is a good way to validate the new interface and implementation.

wwilson commented:
Thanks! 

Meanwhile, what can we do to get this reviewed and merged? It's been 3 months...

roberth commented:
> what can we do to get this reviewed and merged? It's been 3 months...

I know, right? I've advertised this PR on multiple occasions, but so far, no-one has given a review.
A month ago, I've assumed it was because of the size, but #124875 has seen no love either.

wwilson commented:
I saw that https://github.com/NixOS/nixpkgs/pull/124875 got merged! That’s great!!

What are the steps to getting this one reviewed and merged as well?

roberth commented:
I'm too busy to champion this PR :(
Some things that help:

 - Use it in practice to validate the design
   - You can `let inherit (import (pr + "/lib")) sources` where `pr` is a pin of this PR's branch.
 - Show your interest and approval
   - :+1: reaction on the PR top
   - Review the PR (can also be a partial review afaic)
 - Promote it in the community

roberth commented:
This seems to be a candidate for https://github.com/NixOS/rfcs/pull/82

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/friday-hackday-project-nix-filter/12648/6

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-generate-documentation-from-arbitrary-nix-code/22292/8

infinisil reviewed (changes requested):
Taking a closer look at the PR, the biggest problem is how hard it is to understand what the functions `extend`, `empty`, `focusAt`, `cutAt`, `getOriginalFocusPath`, `getSubpath` and `reparent` do. It's neither obvious from the name or their documentation. I'm also not understanding use cases for these, when would I want to use one or the other? I'm giving this a better try here:

## An overview of the functions

The core concept of this PR is that it works with two anchor points of filtered sources:
- root anchor: The root of the store path import, only files under this anchor can be included in the filter
- subpath anchor: The subpath of the root anchor where the filtered source is pointing to. This is what the filtered source evaluates to with `"${source}"` or `source.outPath`

I'm going to use a new notation for a filtered source, where `//` indicates the root anchor:
```nix
/my/project//some/subpath/
            ^            ^
    root anchor      subpath anchor

# When imported into the store:
/nix/store/<hash>-<name>/some/subpath/
```

Using this format, we can explain better what individual functions do to these two anchors and the filtering. Note that these examples are only illustrative, this is not an exact description.
- `focusAt`: Sets the subpath anchor without changing the filtering:
  ```nix
  focusAt ./some /my/project//some/subpath
  -> /my/project//some

  focusAt ./some/subpath/foo /my/project//some/subpath
  -> /my/project//some/subpath/foo
  ```

- `cutAt`: Moves the root anchor down, preventing imports of files outside of the new root:
  ```nix
  cutAt ./some /my/project//some/subpath
  -> /my/project/some//subpath
  ```

- `reparent`: Moves the root anchor up, without adding any new files:
  ```nix
  reparent /my /my/project//some/subpath
  -> /my//project/some/subpath
  ```

- `extend`: Adds new files, moves the root anchor up if necessary:
  ```nix
  extend /my/project//some/subpath [ /my/otherproject/some/file ]
  -> /my//project/some/subpath
  ```

- `empty`: Returns a filtered source with a specific root anchor, but with all files filtered out:
  ```nix
  empty /my/project
  -> /my/project//
  ```

- `getOriginalFocusPath`: Returns the subpath anchor:
  ```nix
  getOriginalFocusPath /my/project//some/subpath
  -> /my/project/some/subpath
  ```

- `getSubpath`: Returns the subpath anchor relative to the root anchor. This is what gets appended to the store path result of the underlying `builtins.filterSource` call:
  ```nix
  getSubpath /my/project//some/subpath
  -> some/subpath
  ```

- `.origSrc` (not really exposed): Returns the root anchor. This is the `src` argument to the underlying `builtins.filterSource` call:
  ```nix
  (/my/project//some/subpath).origSrc
  -> /my/project
  ```

## An alternative model

While I think above gives a nice overview of these functions and potentially makes it easier to give them better names, I'd also like to propose a hopefully simpler to understand alternative set of functions. Notably omitting:
- A way to move the root anchor up (`reparent`, part of `extend`). Users should declare the root anchor manually. This is what everybody is used to already and should only be necessary once per project repository.
- An identity source (`empty`), because it's not necessary if the API is designed to take lists of sources.

- `union :: Path -> ListOf SourceLike -> SourceLike`: Filters a path to only include the directories/files from the given list. It throws an error if any items have a subpath set. This is a simpler `extend`.
- `setSubpath :: Path -> SourceLike -> SourceLike`: Sets the subpath, like doing appending or removing `/path/entries` to the source path, essentially path manipulation. We just can't use `dirOf` and `+` because those don't propagate the filtering function. This is the same as the previous `focusAt`. Note that a `getSubpath` is needed too.
- `limitToSubpath :: SourceLike -> SourceLike`: Limits the files that get imported into the store to the the ones under the subpath set with `setSubpath`, the result has an empty `subpath`. This is like reimporting the path the source is pointing at. This is a simpler `cutAt`.
- Functions to get the `

Notably I think these are easier to separate: `union` can be introduced in one PR, `setSubpath` along with the `mkDerivation` change in another, and `limitToSubpath` in another one (though it depends on `setSubpath`)

roberth commented:
Thanks for the review.

The overview looks good.
It will be nice to cut it down a bit :scissors: 
<sub>`cutAt`</sub> :speak_no_evil: 

Anyway, these are my thoughts now. I'll make some of the changes later.

> While I think above gives a nice overview of these functions and potentially makes it easier to give them better names, I'd also like to propose a hopefully simpler to understand alternative set of functions. Notably omitting:
> 
> * A way to move the root anchor up (`reparent`, part of `extend`). Users should declare the root anchor manually. This is what everybody is used to already and should only be necessary once per project repository.
> * An identity source (`empty`), because it's not necessary if the API is designed to take lists of sources.

`empty` shouldn't be necessary, even with `extend` instead of `union`; same for `reparent`. I probably added them for completeness, but they're a distraction.

> * `union :: Path -> ListOf SourceLike -> SourceLike`: Filters a path to only include the directories/files from the given list. It throws an error if any items have a subpath set. This is a simpler `extend`.

Wait, is this still for omission? I think we're switching to addition now.

Note to self: I do still mention `union` in some doc.

Throwing is a usability and design smell. These were intended to be combinators, so functions to operate on _any_ source input.

What is the `Path` argument for? Is it the source that provides the focus (subpath) like in my earlier iterations that had `union`?

> * `setSubpath :: Path -> SourceLike -> SourceLike`: Sets the subpath, like doing appending or removing `/path/entries` to the source path, essentially path manipulation. We just can't use `dirOf` and `+` because those don't propagate the filtering function. This is the same as the previous `focusAt`.

Correct. I'm ok with this name.

> Note that a `getSubpath` is needed too.

Already added.

> * `limitToSubpath :: SourceLike -> SourceLike`: Limits the files that get imported into the store to the the ones under the subpath set with `setSubpath`, the result has an empty `subpath`. This is like reimporting the path the source is pointing at. This is a simpler `cutAt`.

:+1:

`cutAt` was more complicated than necessary. `limitToSubpath` is a simpler, equally capable replacement.

> * Functions to get the `

Looks like you accidentally your comment :confused: 

> Notably I think these are easier to separate: `union` can be introduced in one PR, `setSubpath` along with the `mkDerivation` change in another, and `limitToSubpath` in another one (though it depends on `setSubpath`)

The subpath features don't rely on `extend` either, so doing those is still a good first step, regardless of "merge" function design.

infinisil commented:
> > * `union :: Path -> ListOf SourceLike -> SourceLike`: Filters a path to only include the directories/files from the given list. It throws an error if any items have a subpath set. This is a simpler `extend`.
> 
> Wait, is this still for omission? I think we're switching to addition now.

I was thinking of removing `extend` and replacing it with `union` here, because ignoring the subpath of `a`, any `extend a bs` can be replaced with `union commonPrefixPath ([ a ] ++ bs)` (with the user manually figuring out `commonPrefixPath`).

> Throwing is a usability and design smell. These were intended to be combinators, so functions to operate on _any_ source input.

As far as I can see, `extend` currently just ignores any subpaths on its list argument, it only takes the subpath from the first argument right? My proposed `union` just changes that to a `throw` when the list argument has any subpaths, because ignoring them would leave people wondering.

If somebody needs a subpath of the result, they can do `setSubpath ./some/path (union ./. [ ... ])`.

> What is the `Path` argument for? Is it the source that provides the focus (subpath) like in my earlier iterations that had `union`?

The path argument of `union` is the root of the store import. It's notably only a path, not a source filter, so no `subpath` or filtering, it only provides the root anchor.

> > * Functions to get the `
> 
> Looks like you accidentally your comment confused

Ah whoops, was gonna say "Functions to get the `subpath`, `origSrc`, etc.", probably a good idea to have that

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/advice-wanted-writing-a-build-system-in-nix/23460/4

infinisil commented:
Recently I worked on an improved set of combinators inspired by this PR. It's currently a [draft Nixpkgs PR](https://github.com/NixOS/nixpkgs/pull/188301#222891), but it needs more eyes! So if you're interested in this, please take a look, try it out, and give feedback! The best overview is here: https://discourse.nixos.org/t/easy-source-filtering-with-file-sets/29117

infinisil commented:
The `lib.fileset` library is now fairly usable, see https://github.com/NixOS/nixpkgs/issues/266356 for the status and updates! As such I don't think we need this PR here anymore, I'll close it. If you use the file set library but run into problems, please report this in https://github.com/NixOS/nixpkgs/issues/266356 (ideally with an issue/PR).
